src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentAction.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Individual action from RenameDocument APIs in <see cref="Renamer"/>.
+        /// 
+        /// See <see cref="RenameDocumentActionSet" />
+        /// </summary>
+        public abstract class RenameDocumentAction
+        {
+            private readonly ImmutableArray<ErrorResource> _errorStringKeys;
+            protected OptionSet OptionSet { get; }
+
+            public virtual ImmutableArray<string> GetErrors(CultureInfo? culture = null)
+                => _errorStringKeys.SelectAsArray(s => string.Format(WorkspacesResources.ResourceManager.GetString(s.FormatString, culture ?? WorkspacesResources.Culture)!, s.Arguments));
+
+            public abstract string GetDescription(CultureInfo? culture = null);
+            internal abstract Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken);
+
+            internal RenameDocumentAction(ImmutableArray<ErrorResource> errors, OptionSet optionSet)
+            {
+                OptionSet = optionSet;
+                _errorStringKeys = errors;
+            }
+
+            internal readonly struct ErrorResource
```

Comment:
how is this different than a Diagnostic?

Reply:
should we just be using Diagnostic in our API?

Reply:
This is an internal only representation and has no real surface area. Tying it to Diagnostic just means that as we make diagnostic changes they would have to update this somewhat unrelated code.

Reply:
Hrmm... i'm not sure i understand.  It's a localizable error that can be used to get user-visible descriptions and whatnot.  Isn't that exactly what a Diagnostic is?  Isn't it really suitable for this domain?  

i.e. you can't do this because these Diagnostic reasons...

Reply:
Diagnostic requires location and is intended to be displayed in the editor. These may not have a location (currently they don't) and will not be displayed in the editor. 

It really was only because I didn't want to write a tuple everywhere...

Reply:
THey don't.  there's Location.NoLocation :)   But i won't belabor the point.  If you're really opposed to it, i'm not going to push it.

Reply:
since this is not public, i'm ok with this.  but i would not be ok if we needed to expose this.